llpc/translator/lib/SPIRV/SPIRVReader.cpp

```diff
@@ -8308,6 +8392,154 @@ llvm::Function *SPIRVToLLVM::createLibraryEntryFunc() {
+// br i1 %shouldLoad, label %loadBlock, label %finalBlock
+//
+// loadBlock:
+// %load = load <4 x float>, <4 x float> ... localArray[runtimeIndex][runtimeIndex * 2]
+// br label %finalBlock
+//
+// finalBlock:
+// %finalResult = phi ... <4 x float> [ zeroinitializer, %entry], [ %load, loadBlock ]
+// ... use %finalResult subsequently
+//
+SPIRVToLLVM::ScratchBoundsCheckInfo SPIRVToLLVM::prepareScratchBoundsCheck(SPIRVValue *toBeWrapped) {
+  constexpr ScratchBoundsCheckInfo nullValue{};
+  if (!getPipelineOptions()->enableScratchAccessBoundsChecks) {
+    return nullValue;
+  }
+
+  if (!toBeWrapped || toBeWrapped->getOpCode() != OpAccessChain) {
+    return nullValue;
+  }
+
+  SPIRVStorageClassKind pointerStorageClass = toBeWrapped->getType()->getPointerStorageClass();
+
+  // We don't want bounds checks if we are outside private or function storage class.
+  if (pointerStorageClass != StorageClassPrivate && pointerStorageClass != StorageClassFunction) {
+    return nullValue;
+  }
+
+  // Fetch the access chain behind the pointer the memory operation uses.
+  const auto accessChainBase = static_cast<SPIRVAccessChainBase *const>(toBeWrapped);
+
+  llvm::SmallVector<SPIRVValue *> indicesVec;
+
+  // Find any integer index, either for accessing the array or a struct member.
+  for (auto *const operand : accessChainBase->getIndices()) {
+    assert(operand->getType()->isTypeInt());
+    indicesVec.push_back(operand);
+  }
+
+  if (indicesVec.empty()) {
+    return nullValue;
+  }
+
+  ArrayRef<SPIRVValue *> indices = indicesVec;
+  SPIRVValue *currentIndex = nullptr;
+
+  auto consumeFirstIndex = [&currentIndex, &indices]() {
+    assert(!indices.empty());
+    currentIndex = indices[0];
+    indices = indices.drop_front();
+  };
+
+  SPIRVType *baseType = accessChainBase->getBase()->getType()->getPointerElementType();
+  consumeFirstIndex();
+
+  // This will contain our final compare result.
+  Value *finalCmpResult = nullptr;
+
+  BasicBlock *entryBlock = getBuilder()->GetInsertBlock();
+  Function *parent = entryBlock->getParent();
+
+  // Build a chain of OOB checks.
+  while (currentIndex) {
+    // Handle either arrays or (nested) structs.
+    if (baseType->getOpCode() == OpTypeArray) {
+      if (const auto baseArray = static_cast<const SPIRVTypeArray *>(baseType)) {
+        // Add the bounds check.
+        if (isConstantOpCode(currentIndex->getOpCode())) {
+          if (static_cast<SPIRVConstant *>(currentIndex)->getZExtIntValue() < baseArray->getArrayLength()) {
+            continue;
+          }
+        }
+
+        auto cmpResult = getBuilder()->CreateICmpULT(
+            transValue(currentIndex, getBuilder()->GetInsertBlock()->getParent(), getBuilder()->GetInsertBlock()),
+            getBuilder()->getInt32(baseArray->getArrayLength()));
+
+        if (finalCmpResult) {
+          finalCmpResult = getBuilder()->CreateAnd(finalCmpResult, cmpResult);
+        } else {
+          finalCmpResult = cmpResult;
```

Comment:
Could you start with `finalCmpResult = TrueVal`/`FalseVal;`? This way you won't have a special case for the initial `And`.

Reply:
Doing so would require adding additional state for checking if an OOB compare was actually inserted (check the `if (finalCmpResult)` at the end of the method). Also, the outcoming LLVM IR will be a little bit ugly, something in the lines of:

```
  %22 = icmp ult i32 %13, 5
  %23 = and i1 true, %22
  %24 = icmp ult i32 %14, 10
  %25 = and i1 %23, %24
  br i1 %25, label %43, label %44
```

Would you mind just leaving it like this?

Reply:
Thanks for the context, sounds good to me.

Reply:
Thanks. I'll leave it like it is then.